### DISCUSSION TITLE
Is there a plan to support asynchronous custom parser api?

### DISCUSSION THREAD

**Discussion by kdy1** [eslint/eslint#14733] [Ideas]
Hi, I'm author of the swc project. https://swc.rs/

The swc proejct has a fast javascript / typescript parser written in rust. It's a native module and as a result it can use the thread pool of libuv.

If eslint supports asynchronous parser api, we can offload parsing to backgorund threads. Parsing will not block the js thread, which is busy processing eslint rules. So performance will be greatly improved.

I'm asking this because  https://eslint.org/docs/developer-guide/working-with-custom-parsers#working-with-custom-parsers does not mention asychronous parser apis.

**Comment by nzakas** [comment]
We don’t support async parser APIs and we don’t have any plans to do so. The low-level Linter class assumes synchronous parsing, so that would have to be rewritten in order to enable this. I’m not saying we wouldn’t ever do that but we don’t have it on the roadmap.

**Comment by ild0tt0re** [comment]
It seems to be at least among the ideas 😍
https://github.com/eslint/eslint/issues/15383

**Reply by nzakas** [reply]
It is something I want to explore further, yes. We’ve received a couple of requests for this so we need to investigate further.

**Reply by kdy1** [reply]
`swc` can emit estree ast (only from rust code), but maybe the performance of `JSON.parse` can be a bottleneck.
https://github.com/swc-project/swc/issues/2175
